### PR TITLE
[Mechanic] Flotsam Drop Rate Attribute

### PR DESCRIPTION
**Pull request by Azure3141**
-----------------------
**Feature:** This PR implements the feature request detailed and discussed in the comments of PR #7746, specifically by @Quantumshark.

## Feature Details
This adds a new attribute, `flotsam chance`, which determines the chance of an outfit becoming flotsam when the ship it is installed in is destroyed.

Before this PR, by default any installed Ammunition outfits would have a 5% chance of being dropped upon ship destruction. This behavior is retained, but specifying the `flotsam chance` on an Ammunition outfit will override these default odds.

## Usage Examples
For example, take this outfit I totally made just now and which is definitely not part of any other content I am working on

```outfit "Fractal Capacitor Cell"
	category "Power"
	cost 200000
	thumbnail "outfit/supercapacitor"
	"mass" 1
	"outfit space" -3
	"energy capacity" 1800
	"ion protection" -0.003
	"flotsam chance" 0.5
```
On ship destruction, every one of these outfits installed on a ship would have a 50% chance of being dropped as flotsam

## Testing Done
I tested outfits with `flotsam chance` values of 0.5 and 1, verifying that they were dropping properly. I also tested Ammunition outfits without the attribute to verify that the previous default behavior was unchanged, and that they dropped at a 5% rate.

## Performance Impact
N/A

### CODE REVIEW

**Comment by RisingLeaf** [comment]
fix this title

**Review by RisingLeaf — APPROVED** [review]
is this tested though?

**Comment by RisingLeaf** [comment]
> > is this tested though?
> 
> Yes.
> 
> > I tested outfits with `flotsam chance` values of 0.5 and 1, verifying that they were dropping properly. I also tested Ammunition outfits without the attribute to verify that the previous default behavior was unchanged, and that they dropped at a 5% rate.

whoops, should read more closely

File: source/Ship.cpp [hunk 1]
```diff
@@ -1586,10 +1586,14 @@ void Ship::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
 					Jettison(it.first, Random::Binomial(it.second, .25));
 				for(const auto &it : cargo.Outfits())
 					Jettison(it.first, Random::Binomial(it.second, .25));
-				// Ammunition has a 5% chance to survive as flotsam
+				// Ammunition has a default 5% chance to survive as flotsam.
 				for(const auto &it : outfits)
-					if(it.first->Category() == "Ammunition")
+				{
+					if(it.first->Get("flotsam chance") > 0.)
+						Jettison(it.first, Random::Binomial(it.second, it.first->Get("flotsam chance")));
```
Comment (Hurleveur) [inline source/Ship.cpp:1593]:
```suggestion
					double flotsamChance = it.first->Get("flotsam chance");
					if(flotsamChance)
						Jettison(it.first, Random::Binomial(it.second, flotsamChance));
```
this attribute cant be negative so no need to check for being higher than 0 being different than it is enough
also some caching

**Review by Hurleveur — APPROVED** [review]

**Review by warp-core — COMMENTED** [review]
It would be nice to be able to have an ammunition outfit with no chance off being dropped as flotsam. I can think of two ways of doing this: provide a way to check if an attribute exists in the outfit attributes Dictionary, so that explicitly setting the attribute to 0 can be differentiated from not having set it at all, or allow negative values and take those to mean no chance, with 0 meaning the default (5% for ammunition, nothing for everything else). I think I prefer the second option, but I suppose it's not super necessary right now.

**Comment by Hurleveur** [comment]
I think its simpler to allow negative values since all attributes default to 0 but that's not really absolutely required for this PR, we could keep it for another one?

**Comment by quyykk** [comment]
We could also remove the special drop rate of ammunition and specify the drop rate on the ammo outfits instead. For backwards compatibility we set the current drop rate for ammo if it isn't set with a helpful deprecation notice.

**Comment by Hurleveur** [comment]
That sounds best to me

**Comment by Amazinite** [comment]
> We could also remove the special drop rate of ammunition and specify the drop rate on the ammo outfits instead. For backwards compatibility we set the current drop rate for ammo if it isn't set with a helpful deprecation notice.

We'd need explicit zeros in either case though. If someone has an ammo outfit that we don't want to have drop, we don't want that to throw a deprecation notice.

**Comment by quyykk** [comment]
we'd use the existing pattern no? A bool like `hasDropRate` in Load.

**Comment by Amazinite** [comment]
Yes, but if we need the same thing in either case, why remove the existing behavior for ammunition?

File: source/Ship.cpp [hunk 2]
```diff
@@ -1586,10 +1586,15 @@ void Ship::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
...
 					Jettison(it.first, Random::Binomial(it.second, .25));
-				// Ammunition has a 5% chance to survive as flotsam
+				// Ammunition has a default 5% chance to survive as flotsam.
 				for(const auto &it : outfits)
-					if(it.first->Category() == "Ammunition")
+				{
+					double flotsamChance = it.first->Get("flotsam chance");
+					if(flotsamChance)
+						Jettison(it.first, Random::Binomial(it.second, flotsamChance));
+					else if(it.first->Category() == "Ammunition")
```
Comment (warp-core) [inline source/Ship.cpp:1595]: You'll also need to update the MINIMUM_OVERRIDES map in the namespace in Outfit.cpp if you do this.
```suggestion
					double flotsamChance = it.first->Get("flotsam chance");
					if(flotsamChance > 0.)
						Jettison(it.first, Random::Binomial(it.second, flotsamChance));
					// 0 valued 'flotsamChance' means default, which is 5% for ammunition.
					// At this point, negative values are the only non-zero values possible.
					// Negative values override the default chance for ammunition
					// so the outfit cannot be dropped as flotsam.
					else if(it.first->Category() == "Ammunition" && !flotsamChance)
```

**Pull request merged**